### PR TITLE
flatMapIterable and flatMapPublisher

### PR DESCRIPTION
**Pull request by wyang14**
Hi @johnmcclean-aol, I have created a new branch off master and implemented flatMapIterable and flatMapPublisher for issue #237 . Please review this PR when you have time. If there is still any issue in the implementation, please let me know.

Thanks.

### CODE REVIEW

File: src/main/java/com/aol/cyclops/control/Maybe.java [hunk 1]
```diff
@@ -72,591 +73,674 @@ public void odd() {
...
+		extends MonadicValue1<T>, Supplier<T>, ConvertableFunctor<T>, Filterable<T>, ApplicativeFunctor<T>, Matchable.ValueAndOptionalMatcher<T> {
+
+	final static Maybe EMPTY = new Nothing<>();
+
+	static <T> Maybe<T> none() {
+		return EMPTY;
+	}
+
+	@Override
+	default <R> Maybe<R> flatMapIterable(Function<? super T, ? extends Iterable<? extends R>> mapper) {
```
Comment (johnmcclean-aol) [inline src/main/java/com/aol/cyclops/control/Maybe.java]: Same comment about Javadoc (I'm working on getting all these methods to have Javadoc and then on the core types will add in more detailed documentation there too).

File: src/main/java/com/aol/cyclops/types/MonadicValue2.java [hunk 2]
```diff
@@ -62,4 +66,25 @@
      * @see com.aol.cyclops.types.MonadicValue#unit(java.lang.Object)
      */
     <T2> MonadicValue2<T1, T2> unit(T2 unit);
+    
+	default <R> MonadicValue2<T1, R> flatMapIterable(Function<? super T2, ? extends Iterable<? extends R>> mapper) {
```
Comment (johnmcclean-aol) [inline src/main/java/com/aol/cyclops/types/MonadicValue2.java]: Could we get more detailed Javadoc here that explains we only take the first entry if there is more than 1 present?

(Then all the sub-classes can reference this definition too).

Thanks!

File: src/main/java/com/aol/cyclops/types/MonadicValue2.java [hunk 3]
```diff
@@ -62,4 +66,25 @@
...
+			if (it.hasNext()) {
+				R r = it.next();
+				return unit(r);
+			} else {
+				return null;
+			}
+		});
+	}
+
+	default <R> MonadicValue2<T1, R> flatMapPublisher(Function<? super T2, ? extends Publisher<? extends R>> mapper) {
```
Comment (johnmcclean-aol) [inline src/main/java/com/aol/cyclops/types/MonadicValue2.java]: Same commend as for Iterable 

Could we get more detailed Javadoc here that explains we only take the first entry if there is more than 1 present?

(Then all the sub-classes can reference this definition too).

Thanks!

**Comment by johnmcclean-aol** [comment]
Hey Will, 

Looks really good - added some comments about additional Javadoc. Would be really great if you have some time? 

Thanks,
John

**Comment by wyang14** [comment]
Hi @johnmcclean-aol , thanks for your comments. I will add the java docs in the PR asap.

**Comment by wyang14** [comment]
Hi @johnmcclean-aol , I have added comments to the Java classes/interfaces. Please let me know if you see any syntax error in it. Thanks.

**Pull request merged**